modules/fx/arrow-fx/src/main/kotlin/arrow/fx/internal/IQueue.kt

```diff
@@ -0,0 +1,87 @@
+package arrow.fx.internal
+
+import arrow.core.Tuple2
+import arrow.core.Option
+import arrow.core.None
+import arrow.core.Some
+import arrow.core.toT
+import arrow.core.extensions.list.foldable.exists
+import arrow.core.extensions.list.foldable.nonEmpty
+
+/**
+ *  Port of `scala.collection.immutable.Queue`
+ * `Queue` objects implement data structures that allow to
+ *  insert and retrieve elements in a first-in-first-out (FIFO) manner.
+ *
+ *  `Queue` is implemented as a pair of `List`s, one containing the ''in'' elements and the other the ''out'' elements.
+ *  Elements are added to the ''in'' list and removed from the ''out'' list. When the ''out'' list runs dry, the
+ *  queue is pivoted by replacing the ''out'' list by ''in.reverse'', and ''in'' by ''Nil''.
+ *
+ */
+
+class IQueue<A> private constructor(val lIn: List<A>, val lOut: List<A>) {
```

Comment:
Is there any alternative to porting this data structure? Does it need to be immutable?